python/whylogs/api/logger/experimental/logger/actor/process_rolling_logger.py

```diff
@@ -0,0 +1,473 @@
+from whylogs.core.stubs import pd
+
+MessageType = Union[
+    FlushMessage,
+    RawLogMessage,
+    RawLogEmbeddingsMessage,
+    RawPubSubMessage,
+    RawPubSubEmbeddingMessage,
+    LogMessage,
+    CloseMessage,
+    ProcessLoggerStatusMessage,
+]
+
+DataTypes = Union[str, int, float, bool, List[float], List[int], List[str]]
+
+DictType = TypeVar("DictType", bound="Union[LogRequestDict, LogEmbeddingRequestDict]")
+Loggable = Union[pd.DataFrame, Dict[str, Any]]
+
+
+class ProcessRollingLogger(ProcessActor[MessageType], DataLogger[Dict[str, ProcessLoggerStatus]]):
+    def __init__(
+        self,
+        aggregate_by: TimeGranularity = TimeGranularity.Hour,
+        write_schedule: Optional[Schedule] = Schedule(cadence=TimeGranularity.Minute, interval=10),
+        schema: Optional[DatasetSchema] = None,
+        sync_enabled: bool = False,
+        current_time_fn: Optional[Callable[[], int]] = None,
+        queue_config: QueueConfig = QueueConfig(),
+        thread_queue_config: QueueConfig = QueueConfig(),
+        debug_writers: List[Writer] = [],
+    ) -> None:
+        super().__init__(queue_config=queue_config)
+        self._sync_enabled = sync_enabled
+        self._thread_queue_config = thread_queue_config
+        self.current_time_ms = current_time_fn or current_time_ms
+        self.loggers: Dict[str, ThreadRollingLogger] = {}
+        self.debug_writers = debug_writers
+        self.write_schedule = write_schedule
+        self.schema = schema
+        self.aggregate_by = aggregate_by
+        self._pipe_signaler: Optional[PipeSignaler] = PipeSignaler() if sync_enabled else None
+        self._session = get_current_session()
+
+    def create_writers(self, dataset_id: str) -> List[Writer]:
+        return [
+            Writers.get(
+                "whylabs",
+                # TODO get these from the session after that pr is merged
+                # org_id=self.env_vars.whylabs_org_id,
+                # api_key=self.env_vars.whylabs_api_key,
+                dataset_id=dataset_id,
+            )
+        ]
+
+    def _create_logger(self, dataset_id: str) -> ThreadRollingLogger:
+        writers = self.create_writers(dataset_id)
+
+        logger = ThreadRollingLogger(
+            aggregate_by=self.aggregate_by,
+            writers=[*self.debug_writers, *writers],
+            schema=self.schema,
+            write_schedule=self.write_schedule,
+            current_time_fn=self.current_time_ms,
+            queue_config=self._thread_queue_config,
+        )
+
+        self._logger.info(f"Created logger for {dataset_id}")
+        return logger
+
+    def _get_logger(self, dataset_id: str) -> ThreadRollingLogger:
+        if dataset_id not in self.loggers:
+            self.loggers[dataset_id] = self._create_logger(dataset_id)
+        return self.loggers[dataset_id]
+
+    def process_batch(self, batch: List[MessageType], batch_type: Type) -> None:
+        if batch_type == FlushMessage:
```

Comment:
This has to do with the batching logic. There isn't a great way to represent this in the type system though.

The batching logic in the super class guarantees that you'll only get subtypes of a particular message and it explicitly tells the subclass which type it is via the batch_type. I felt that it would be a little to squishy/implicit to just assume that in the sub types and check an arbitrary element's type. There are a few cases that wouldn't cover, like the empty list (no current use case) and cases where messages are subtypes of each other.

And yeah there is no runtime typecheck here from a value, just an equality check, but I wasn't super concerned about perf even if that wasn't true.

This is kind of ugly though.